### PR TITLE
chore: update changelog, with target date for v2.11.0 release

### PR DESCRIPTION
**Pull request by rwgk**
## Description
We have a lot of accumulated changes. Proposed release date: June 2, 2023

Note for easy reference: v2.10.0 was released on Jul 15, 2022.

Please help proof-reading the changelog.

PRs to be merged before the release (**please review or comment by June 2**):

* #4671
* #4674
* #4678
* Any others likely to be ready soon?

## Suggested changelog entry:

```rst

```

### CODE REVIEW

**Comment by henryiii** [comment]
I'll be doing some stuff for Pyodide support, would like to make sure that's working before a release - including a pyodide CI job I've been working on using out-of-tree testing using scikit-build-core (which is really cool, but broken by a bug in emscripten I'm working on). I'm at the scientific-python developer summit all next week, and won't have much time to finish that up till after I expect, so targeting a time after that would be preferable, I think.

I have a feeling 2.10's date was related to SciPy.

Did you use nox to generate the (initial draft) of the changelog update?

File: docs/changelog.rst [hunk 1]
```diff
@@ -40,8 +75,16 @@ Build system improvements:
 * Moved the linting framework over to Ruff.
   `#4483 <https://github.com/pybind/pybind11/pull/4483>`_
 
+* Skip lto checks and target generation when
+  ``CMAKE_INTERPROCEDURAL_OPTIMIZATION`` is defined.
+  `#4643 <https://github.com/pybind/pybind11/pull/4643>`_
+
+* No longer inject -stdlib=libc++, not needed for modern Pythons (macOS 10.9+).
```
Comment (henryiii) [inline docs/changelog.rst:82]:
```suggestion
* No longer inject ``-stdlib=libc++``, not needed for modern Pythons (macOS 10.9+).
```

File: docs/changelog.rst [hunk 2]
```diff
@@ -10,10 +10,24 @@ Changes will be added here periodically from the "Suggested changelog entry"
...
+   in which ``std::is_move_constructible`` does not work as needed. This is
+   very similar to the long-established
+   ``pybind11::detail::is_copy_constructible``.
+  `#4631 <https://github.com/pybind/pybind11/pull/4631>`_
 
-Changes will be summarized here periodically.
+* Introduce recursive_container_traits.
+  `#4623 <https://github.com/pybind/pybind11/pull/4623>`_
+
+* pybind11/type_caster_pyobject_ptr.h was added to support automatic wrapping
```
Comment (henryiii) [inline docs/changelog.rst:27]:
```suggestion
* ``pybind11/type_caster_pyobject_ptr.h`` was added to support automatic wrapping
```

File: docs/changelog.rst [hunk 3]
```diff
@@ -10,10 +10,24 @@ Changes will be added here periodically from the "Suggested changelog entry"
...
+New features:
+
+* ``pybind11::detail::is_move_constructible`` can now be specialized for cases
+   in which ``std::is_move_constructible`` does not work as needed. This is
+   very similar to the long-established
+   ``pybind11::detail::is_copy_constructible``.
+  `#4631 <https://github.com/pybind/pybind11/pull/4631>`_
 
-Changes will be summarized here periodically.
+* Introduce recursive_container_traits.
```
Comment (henryiii) [inline docs/changelog.rst:24]:
```suggestion
* Introduce ``recursive_container_traits``.
```

File: docs/changelog.rst [hunk 4]
```diff
@@ -31,6 +45,27 @@ Changes:
   sizes slightly (~1.5%) but the error messages are much more informative.
   `#4463 <https://github.com/pybind/pybind11/pull/4463>`_
 
+* Setter return values (which are inaccessible for all practical purposes) are
+  no longer converted to Python (only to be discarded).
+  `#4621 <https://github.com/pybind/pybind11/pull/4621>`_
+
+* Allow lambda specified to function definition to be noexcept(true) in C++17.
```
Comment (henryiii) [inline docs/changelog.rst:52]:
```suggestion
* Allow lambda specified to function definition to be ``noexcept(true)`` in C++17.
```

Comment (rwgk) [reply docs/changelog.rst]: Applied, thanks!

**Comment by rwgk** [comment]
> so targeting a time after that would be preferable, I think.

I changed the target date to June 2, do you think that could work?

I don't have a direct interest in releases (I work against latest master everywhere), but the thought that people are spending hours debugging GIL issues unnecessarily makes me a bit sad; #4246 is still only on master, but in no release. I believe exactly the people who're likely to benefit the most, those with limited experience, will also be the ones least likely to work against master. 

> Did you use nox to generate the (initial draft) of the changelog update?

I looked, but then decided to simply try `python3 tools/make_changelog.py` and when that worked I didn't look any further. Thanks for the script btw, it eliminates a lot of the grunge work.

**Comment by rwgk** [comment]
I think the v2.11.0 release could also be a good opportunity to purge Python 3.6 support. End of life was already 23 Dec 2021, 1 year 4 months ago. That would get rid of all `WITH_THREAD` (see below).

I could do the work (I think it's only a couple hours), but

* can someone think of reasons to continue Python 3.6 support?

* Could someone please help reviewing with quick turnarounds (~24 hours)?

```
include/pybind11/embed.h:#    if defined(WITH_THREAD) && PY_VERSION_HEX < 0x03070000
include/pybind11/gil.h:#if defined(WITH_THREAD) && !defined(PYBIND11_SIMPLE_GIL_MANAGEMENT)
include/pybind11/gil.h:#if defined(WITH_THREAD)
include/pybind11/gil.h:#else // WITH_THREAD
include/pybind11/gil.h:#endif // WITH_THREAD
include/pybind11/detail/type_caster_base.h:#if defined(WITH_THREAD)
include/pybind11/detail/common.h:// Reminder: WITH_THREAD is always defined if PY_VERSION_HEX >= 0x03070000
include/pybind11/detail/internals.h:#if defined(WITH_THREAD) && defined(PYBIND11_SIMPLE_GIL_MANAGEMENT)
include/pybind11/detail/internals.h:#if defined(WITH_THREAD)
include/pybind11/detail/internals.h:#    if defined(WITH_THREAD)
include/pybind11/detail/internals.h:#if defined(WITH_THREAD)
include/pybind11/detail/internals.h:#if defined(WITH_THREAD)
include/pybind11/detail/internals.h:#if defined(WITH_THREAD) && PYBIND11_INTERNALS_VERSION == 4
include/pybind11/detail/internals.h:#endif //  defined(WITH_THREAD) && PYBIND11_INTERNALS_VERSION == 4
tests/test_call_policies.cpp:#if defined(WITH_THREAD) && !defined(PYPY_VERSION)
```

**Comment by rwgk** [comment]
> Since 2.11 is almost out and all the work is already there to keep in on 3.6, would it be better to drop 3.6 right after? No strong reason to keep supporting imo, only manylinux is left still supporting 3.6.

Sounds good. That way it'll live ~1 year longer in the latest release series, but won't bother us anymore on master.

> Could you drop the Version number change them merge?

Dropped and ready to merge (I'll change the title first).

> That way the labels can be cleared

Do we clear them manually?

> First beta of 3.12 is out soon!

With cpython main from a few hours ago the pybind11 unit tests only build, pytest dies at startup: https://github.com/python/cpython/pull/103912#issuecomment-1555990333

**Comment by rwgk** [comment]
> First beta of 3.12 is out soon!

It is out now, with two bugs I knew about ahead of time, which means we cannot do much with 3.12 beta 1 :-(

One bug is fixed already (even before the cut but it was too late anyway).

The other I/we still have to figure out.

**Comment by rwgk** [comment]
> > That way the labels can be cleared
> 
> Do we clear them manually?

Easy enough: done.

**Pull request merged**